sdk/dslink/src/main/java/org/dsa/iot/dslink/link/SubscriptionHelper.java

```diff
@@ -0,0 +1,192 @@
+    public void clear() {
+        if (subscriptions != null) {
+            subscriptions.clear();
+        }
+    }
+
+    /**
+     * Clears a single subscription without calling unsubscribe on the requester.
+     */
+    public SubscriptionHelper clear(String path) {
+        Adapter a = subscriptions.remove(path);
+        if (a != null) {
+            requester.unsubscribe(path, null);
+        }
+        return this;
+    }
+
+    /**
+     * Safely subscribes the path, even if it is already subscribed.  Calls
+     * subscribe with a SubData for the given path and a QOS of 0.
+     */
+    public SubscriptionHelper subscribe(String path,
+                                        Handler<SubscriptionValue> handler) {
+        return subscribe(new SubData(path, 0), handler);
+    }
+
+    /**
+     * Safely subscribes the path, even if it is already subscribed.  The path will
+     * be subscribed at the give QOS of the very first subscription.  Subsequent
+     * subscriptions will get the same QOS.
+     *
+     * @param path The first scription
+     */
+    @SuppressFBWarnings("AT_OPERATION_SEQUENCE_ON_CONCURRENT_ABSTRACTION")
+    public SubscriptionHelper subscribe(SubData path,
+                                        Handler<SubscriptionValue> handler) {
+        Adapter a = subscriptions.get(path.getPath());
+        if (a == null) {
+            Adapter adapter = new Adapter();
+            adapter.add(handler);
+            requester.subscribe(path, adapter);
+            subscriptions.put(path.getPath(), adapter);
+        } else {
+            a.add(handler);
+        }
+        return this;
+    }
+
+    /**
+     * Only unsubscribes the given handler.
+     */
+    @SuppressFBWarnings("AT_OPERATION_SEQUENCE_ON_CONCURRENT_ABSTRACTION")
+    public SubscriptionHelper unsubscribe(String path,
+                                          Handler<SubscriptionValue> handler) {
+        Adapter a = subscriptions.get(path);
+        if (a != null) {
+            a.remove(handler);
+            if (a.size() == 0) {
+                requester.unsubscribe(path, null);
+                subscriptions.remove(path);
+            }
+        }
+        return this;
+    }
+
+    /**
+     * Only unsubscribes the given handler.
+     */
+    public SubscriptionHelper unsubscribe(SubData path,
+                                          Handler<SubscriptionValue> handler) {
+        return unsubscribe(path.getPath(),handler);
+    }
+
+    /**
+     * Unsubscribe all handlers for the given path.
+     */
+    public SubscriptionHelper unsubscribeAll(String path) {
+        Adapter a = subscriptions.remove(path);
+        if (a != null) {
+            requester.unsubscribe(path, null);
+        }
+        return this;
+    }
+
+
+    // Inner Classes
+    // -------------
+
+    /**
+     * Only identical instances are equal (==).
+     */
+    @SuppressFBWarnings("SE_COMPARATOR_SHOULD_BE_SERIALIZABLE")
+    private static class HandlerComparator
```

Comment:
My understanding is static classes offer better performance in some way.  Not sure how, maybe because they don't a this pointer back to the outer class or something...